source/engine/physics2/jolt/ConstraintManager.cpp

```diff
@@ -0,0 +1,219 @@
+#include "ConstraintManager.h"
+
+#include "Jolt/Jolt.h"
+#include "Jolt/Physics/PhysicsSystem.h"
+#include "Jolt/Physics/Constraints/TwoBodyConstraint.h"
+
+namespace nc::physics
+{
+auto ConstraintManager::AddConstraint(const ConstraintInfo& createInfo,
+                                      Entity owner,
+                                      JPH::Body* ownerBody,
+                                      Entity target,
+                                      JPH::Body* targetBody) -> Constraint&
+{
+    auto handle = m_factory.MakeConstraint(createInfo, *ownerBody, *targetBody);
+    m_physicsSystem->AddConstraint(handle);
+
+    const auto ownerId = owner.Index();
+    const auto targetId = target.Index();
+    const auto index = [&]()
+    {
+        if (m_freeIndices.empty())
+        {
+            const auto i = static_cast<uint32_t>(m_handles.size());
+            m_handles.push_back(handle);
+            m_pairs.emplace_back(ownerId, targetId);
+            return i;
+        }
+
+        const auto i = m_freeIndices.back();
+        m_freeIndices.pop_back();
+        m_handles[i] = handle;
+        m_pairs[i] = ConstraintPair{ownerId, targetId};
+        return i;
+    }();
+
+    auto& ownerState = TrackConstraint(ownerId, index);
+    auto& constraint = ownerState.constraints.emplace_back(createInfo, target, index);
+    if (target.Valid())
+    {
+        TrackConstraint(targetId, index);
+    }
+
+    return constraint;
+}
+
+void ConstraintManager::EnableConstraint(Constraint& constraint, bool enabled)
+{
+    constraint.m_enabled = enabled;
+    GetConstraintHandle(constraint.GetId())->SetEnabled(enabled);
+}
+
+void ConstraintManager::UpdateConstraint(Constraint& constraint)
+{
+    auto oldHandle = GetConstraintHandle(constraint.GetId());
+    auto [ownerBody, targetBody] = GetConstraintBodies(oldHandle);
+    ReplaceInternalConstraint(constraint, oldHandle, ownerBody, targetBody);
+}
+
+void ConstraintManager::UpdateConstraintTarget(Constraint& constraint,
+                                               Entity referenced,
+                                               JPH::Body* referencedBody)
+{
+    constraint.m_otherBody = referenced;
+
+    const auto id = constraint.GetId();
+    auto oldHandle = GetConstraintHandle(id);
+    auto [ownerBody, _] = GetConstraintBodies(oldHandle);
+    ReplaceInternalConstraint(constraint, oldHandle, ownerBody, referencedBody);
+
+    auto& [trackedOwner, trackedReference] = m_pairs.at(id);
+    NC_ASSERT(trackedOwner != referenced.Index(), "Attempt to self assign constraint attachment");
+    if (trackedReference != Entity::NullIndex)
+    {
+        UntrackConstraint(trackedReference, id, false);
+    }
+
+    trackedReference = referenced.Index();
+    if (referenced.Valid())
+    {
+        TrackConstraint(trackedReference, id);
+    }
+}
+
+void ConstraintManager::RemoveConstraint(ConstraintId constraintId)
+{
```

Comment:
There is a lot of state that has to be managed for constraint removal. What happens if any of these methods (like `UntrackConstraint`, for example, throw? I don't have a great suggestion to safeguard against this - but it might be worth adding some additional sanity checks at the beginning to make sure we know we aren't going to throw before we start exchanging and untracking things

Reply:
I can move the the assert if nullptr first always (you'r'e right that in `RemoveConstraint` and `RemoveConstraints` we could end up with nulling the pointer before throwing on a bad id!). As for any other functions throwing later, they could only throw if there's a bug in the tracking, which can't really be left in a valid/recoverable state.